connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java

```diff
@@ -168,32 +171,57 @@ String fullTableName(TableId tableId) {
-      } else if (type == TableDefinition.Type.VIEW
-          || type == TableDefinition.Type.MATERIALIZED_VIEW
-          || (type == TableDefinition.Type.TABLE && filter.isPresent())) {
-        // run a query
-        String table = fullTableName(tableInfo.getTableId());
-        String sql = format("SELECT COUNT(*) from `%s` WHERE %s", table, filter.get());
-        TableResult result = bigQuery.query(QueryJobConfiguration.of(sql));
-        return result.iterateAll().iterator().next().get(0).getLongValue();
-      } else {
-        throw new IllegalArgumentException(
-            format(
-                "Unsupported table type %s for table %s",
-                type, fullTableName(tableInfo.getTableId())));
-      }
-    } catch (InterruptedException e) {
-      throw new BigQueryConnectorException(
-          "Querying table size was interrupted on the client side", e);
+  public TableStatistics calculateTableSize(TableInfo tableInfo, Optional<String> filter) {
+    TableDefinition.Type type = tableInfo.getDefinition().getType();
+    if (type == TableDefinition.Type.TABLE) {
+      // if this is a table, there's no need to query the total number of rows
+      OptionalLong numberOfFilteredRows =
+          toOptionalLong(
+              filter
+                  .map(f -> countWithFilter(tableInfo.getTableId(), f))
+                  .map(row -> row.get(0).getLongValue()));
+      return new TableStatistics(tableInfo.getNumRows().longValue(), numberOfFilteredRows);
+    } else if (type == TableDefinition.Type.VIEW
+        || type == TableDefinition.Type.MATERIALIZED_VIEW) {
+      // For views, we need to calculate the number of rows
+      String table = fullTableName(tableInfo.getTableId());
+      String sql =
+          filter
+              .map(
+                  f ->
+                      format(
+                          "SELECT (SELECT COUNT(*) from `%s`), (SELECT COUNT(*) from `%s` WHERE %s)",
+                          table, table, f))
+              .orElse(format("SELECT  COUNT(*) from `%s`", table));
+      FieldValueList row = queryForSingleRow(sql);
+      long numberOfRows = row.get(0).getLongValue();
+      OptionalLong numberOfFilteredRows =
+          toOptionalLong(filter.map(ignored -> row.get(1).getLongValue()));
+      return new TableStatistics(numberOfRows, numberOfFilteredRows);
+    } else {
+      throw new IllegalArgumentException(
+          format(
+              "Unsupported table type %s for table %s",
+              type, fullTableName(tableInfo.getTableId())));
     }
   }
+
+  private OptionalLong toOptionalLong(Optional<Long> opt) {
+    return opt.map(LongStream::of).orElseGet(LongStream::empty).findFirst();
+  }
+
+  private FieldValueList countWithFilter(TableId tableId, String filter) {
+    String table = fullTableName(tableId);
+    String sql = format("SELECT COUNT(*) from `%s` WHERE %s", table, filter);
+    return queryForSingleRow(sql);
```

Comment:
again, I think this should be opt in.

Reply:
In addition to the opt-in/opt out, it may also make sense to look at the basic table stats and filters to figure out ways to avoid running a query, whenever possible. For instance, if the table is tiny (say well under 10 MB, the Spark default broadcast join threshold), we could just report the full table size. OTOH, if the table is very large and if there are filters on clustering/partition keys, we can estimate the filter ratio.

Reply:
> OTOH, if the table is very large and if there are filters on clustering/partition keys, we can estimate the filter ratio.

I think this could be handled by a dry-run?

Reply:
@emkornfield 

> I think this could be handled by a dry-run?

Unfortunately no

Reply:
Are we targeting these statistics for BHJ conversion? Can we just leverage spark's estimation logic here for filters and aggregates instead of running the query?

Reply:
I believe we are taking a different approach now and trying to use data from the Read API directly.   We might want to close this PR?

Reply:
I agree. Let's reopen when relevant